feat: Added token grant_type in JWT payload

jawad-khan opened the pull request:
## Description
To exchange jwt with session cookies we need to determine JWT grant type in AccessTokenExchangeView.
JWT only having password grant type will be allowed to exchange session.

## Supporting information

JIRA Issue: https://2u-internal.atlassian.net/browse/LEARNER-8886

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-add-grant-type-in-jwt-payload.rst:
```diff
@@ -0,0 +1,27 @@
+Context
+-------
+
+Edx mobile apps are `migrating from Dot access token to JWT token`_ authentication.Therefore gradually adding JWT support in required APIS.
```
1. Can we make this ADR 14, and insert "Mobile migration to JWTs" as ADR 13?
2. I would update this link to point to the new ADR 13, instead of a private Jira ticket.
3. I would avoid using the language "migrating from Dot", because we are still using DOT, so it gets confusing.
4. Should we say "Open edX mobite apps"?

robrap replied:
Thanks for renaming, but can you also insert the requested ADR 13? Ideally, we could copy the Confluence page into an ADR. Worst case, the ADR mostly links to the Confluence page.

robrap replied:
Still awaiting ADR 13.

robrap replied:
Done.

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-add-grant-type-in-jwt-payload.rst:
```diff
@@ -0,0 +1,27 @@
+
+Edx mobile apps are `migrating from Dot access token to JWT token`_ authentication.Therefore gradually adding JWT support in required APIS.
+To exchange JWT with session cookies we need to determine the grant type of the access token first.
+We are already doing this in ``/oauth2/login`` view for dot access token, but we don't have grant type in JWT.
```
What do you think of the following?
```suggestion
As part of the current mobile authentication flow, we exchange an opaque access token for a session This is only allowed for access tokens created with the password grant type. However, as we migrate from using the opaque access token to a JWT, we are no longer able to determine the grant type of the JWT.
```

robrap replied:
This update was used.

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-add-grant-type-in-jwt-payload.rst:
```diff
@@ -0,0 +1,27 @@
+Decisions
+---------
+
+A new key ``grant_type`` will be added in payload while creating JWT. Key ``grant_type`` will be determined through dot access token for mobile flow.
```
```suggestion
A new claim ``grant_type`` will be added to the payload while creating the JWT. The claim ``grant_type`` will be determined using the grant type information stored for the django-oauth-toolkit (DOT) access token that the JWT was based off of during the mobile flow.
```

robrap replied:
Completed.

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-add-grant-type-in-jwt-payload.rst:
```diff
@@ -0,0 +1,27 @@
+---------
+
+A new key ``grant_type`` will be added in payload while creating JWT. Key ``grant_type`` will be determined through dot access token for mobile flow.
+Since we create JWT from access token in mobile flow therefore its value will be ``password``. We can provide any value other than password and by default it will be an empty string.
```
```suggestion
JWTs created from an access token in mobile flow will have a grant type of ``password``, and can be used for the session exchange.

At this time, all values for grant type may not yet be supported, and an empty string will be used by default to represent any grant type that is not password. However, additional grant type values may be implemented as-needed.
```

robrap replied:
Actually, could you manually test the following locally and share what you see in the JWT:
1. Call to access_token using a password.
2. Call to access_token using client credentials.
3. What you see in the JWT payload cookie in the browser.

This ADR note will probably need to be updated accordingly.

jawad-khan replied:
When I called access token I got following grant_types:
with password: "password"
with client credentials: "client-credentials"
In browser jwt cookie I got "password" stored as grant_type.

jawad-khan replied:
BTW you can test these changes on [this sandbox](iap.sandbox.edx.org/).

jawad-khan replied:
> At this time, all values for grant type may not yet be supported, and an empty string will be used by default to represent any grant type that is not password. However, additional grant type values may be implemented as-needed.

This line is not 100% true since we are supporting all the grant types supported by DOT token. and will only have empty string when no grant type is provided.

robrap replied:
1. Note: I see the following in New Relic, but this is from the [POST param](https://github.com/openedx/edx-platform/blob/fe6b666f5b0bea853ba627e7cff4adbab2904a0d/openedx/core/djangoapps/oauth_dispatch/views.py#L105), and not what is in the DB:
```
Oauth Grant Type | Response.Status | Transactions
<BLANK>  | 200 | 919178436
refresh_token | 200 | 2427823
<BLANK>  | 204 | 1792403
<BLANK>  | 201 | 1620052
<BLANK>  | 206 | 755693
client_credentials | 200 | 155905
authorization_code | 200 | 145699
password | 200 | 98034
<BLANK>  | 202 | 1233
```
2. I still don't understand what a blank indicates, and whether this would be `''` for our new `grant_type`? If authentication is successful, doesn't some grant type needed to have been checked?
3. Is there a way to ensure that the JWT will have the right `grant_type` in the newer mobile JWT flow before we switch to it? Maybe this is obvious, but the New Relic data has me concerned. We should work out a dashboard to watch for launch, and make sure we have all the metrics we need.
4. We should remove the last consequence, or update it to clarify when the empty string will be used. It's still unclear to me what it means.

robrap replied:
To clarify, I am not concerned with the rollout of this PR. I am concerned with the final migration of Mobile to a JWT-based workflow, and ensuring there are no unforeseen auth issues that affects successful throughput. This doesn’t affect this PR, or even the other open PR, but just the final cutover. We may end up wanting to add custom attributes to track and compare.

jawad-khan replied:
Currently we are testing these changes in a sandbox environment and flow is working fine. Maybe we can put some extra analytics in sandbox to look for grant type in database. what do you say?

robrap replied:
If sandbox testing is going well, that's great. You are testing with a refresh token as well?

Separately, for final rollout, the question will be what you'll want to compare to see if successful mobile traffic seems comparable before/after. We may have all the data we need. Otherwise, we may want to add some data. I'm happy to help put something together with you at some point. This isn't urgent. Are you planning on using some sort of toggle so you can switch back and forth, in case of issues?

jawad-khan replied:
We will start full testing with all devices and platforms from next week. Currently we just tested flow and it was working fine.
We don't have a toggle for authentication yet but we can send app update in the worst case. Currently we only have a toggle for ecommerce apis.
This looks great and will help us analyze data. I'll surely need your help on this in future.

robrap replied:
Note: I'm considering this resolved. We can take up rollout questions elsewhere.

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-add-grant-type-in-jwt-payload.rst:
```diff
@@ -0,0 +1,27 @@
+Consequences
+------------
+
+* JWT will have an additional attribute i.e. grant_type in its payload, which we can use to take decisions regarding auth exchange.
```
```suggestion
* JWT will have an additional claim "grant_type" in its payload, which we can use to take decisions regarding auth exchange.
* At this time, all values for grant type may not yet be supported. An empty string will be used by default to represent any grant type that is not "password". Additional grant type values may be implemented as required in the future.
```

robrap commented on openedx/core/djangoapps/oauth_dispatch/jwt.py:
```diff
@@ -71,6 +71,7 @@ def create_jwt_token_dict(token_dict, oauth_adapter, use_asymmetric_key=None):
         use_asymmetric_key=use_asymmetric_key,
         is_restricted=oauth_adapter.is_client_restricted(client),
         filters=oauth_adapter.get_authorization_filters(client),
+        grant_type=access_token.application.authorization_grant_type,
```
Is there any risk to this line? I don't think so, and hope not, but am wondering if we need a rollout toggle or anything. I guess we'll see what test coverage looks like.

jawad-khan replied:
Should we do this in try catch?

robrap replied:
Sure. What if you move this to a try catch above, and set grant_type to `''` and `log.exception('Unable to get grant_type from access token.')`. Note: `log.exception` automatically includes exception details and stacktrace.

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/mixins.py:
```diff
@@ -95,6 +95,9 @@ def _decode_jwt(verify_expiration):
             expected['is_restricted'] = should_be_restricted
 
+        if grant_type is not None:
+            expected['grant_type'] = grant_type
```
Should this be?:
```suggestion
        expected['grant_type'] = grant_type or ''
```

robrap replied:
Done.

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/test_jwt.py:
```diff
@@ -22,7 +23,7 @@ def setUp(self):
         self.default_scopes = ['email', 'profile']
 
-    def _create_client(self, oauth_adapter, client_restricted):
+    def _create_client(self, oauth_adapter, client_restricted, grant_type=dot_models.Application.GRANT_PASSWORD):
```
1. I wonder if we should use "password" in place of dot_models.Application.GRANT_PASSWORD. In theory, the `oauth_adapter` could be DOT, or other, although we now only support DOT.
2. Also, in place of "password" as the default, I think it might be more clear to make the default `None`, and to have to include this in the tests that require "password" grant.

jawad-khan replied:
grant_type has NOT NULL constraint so adding '' by default.

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/test_jwt.py:
```diff
@@ -33,14 +34,15 @@ def _create_client(self, oauth_adapter, client_restricted):
         return client
 
-    def _get_token_dict(self, client_restricted, oauth_adapter):
+    def _get_token_dict(self, client_restricted, oauth_adapter, grant_type=dot_models.Application.GRANT_PASSWORD):
```
Same changes requested as in `_create_client`.

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/test_jwt.py:
```diff
@@ -163,3 +165,33 @@ def test_scopes(self):
+
+        self.assert_valid_jwt_access_token(
+            jwt_token_dict["access_token"], self.user, self.default_scopes,
+            grant_type=dot_models.Application.GRANT_PASSWORD,
```
I think I'd drop the import above, and replace `dot_models.Application.GRANT_PASSWORD` with `"password"` everywhere, for reasons I stated above. Let me know if you disagree.

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/test_jwt.py:
```diff
@@ -163,3 +165,33 @@ def test_scopes(self):
+            grant_type=dot_models.Application.GRANT_PASSWORD,
+        )
+
+    def test_empty_str_grant_type(self):
```
1. Maybe it would be better to replace with `test_grant_type_of_none`, and test that `None` => `''`?
2. Additional testing: Can we also test that the views that mobile relies on return the expected claim? I think that may include the third-party token PR, but maybe you can test that locally after bringing in these changes on a new test branch or something. As importantly, we want to make sure that views where we **don't** expect "password", we don't get it.

robrap replied:
@jawad-khan: Can you let me know if you added tests for #2, or tested separately, or what happened. Thanks. I need to run. Maybe this will be obvious when I complete the review anyway.

jawad-khan replied:
Maybe I am not getting your point clear but you can correct me. I did some testing and used other grant types and got expected grant types in return. Also[ this test case](https://github.com/openedx/edx-platform/pull/30519/files#diff-228ea3bbe3ba37a2cade55e78bf7788d911475f911a05a9840a7d5571adfee86R337) ensure this.

robrap reviewed (commented):
Thanks. Happy to discuss any questions if any of this is not clear.

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0014-add-grant-type-in-jwt-payload.rst:
```diff
@@ -0,0 +1,30 @@
+This is only allowed for access tokens created with the password grant type.
+However, as we migrate from using the opaque access token to a JWT, we are no longer able to determine the grant type of the JWT.
+
+.. _migrating from Dot access token to JWT token: 0013-mobile-migration-to jwt.rst
```
1. I'm still looking for this ADR. Did you think it was added?
2. The label needs to change to `migrating to JWT token`.
3. The URL needs to be fixed. The filename has a space in it, and I'm not sure if the filename only will work.

robrap replied:
Done.

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0014-add-grant-type-in-jwt-payload.rst:
```diff
@@ -0,0 +1,30 @@
+Context
+-------
+
+Open edX mobile apps are `migrating to JWT token`_ authentication.Therefore gradually adding JWT support in required APIS.
```
```suggestion
Open edX mobile apps are `migrating to JWT token`_ authentication.Therefore gradually adding JWT support in required APIS.
```
```suggestion
Open edX mobile apps are `migrating to JWT token`_ authentication.
```

robrap replied:
Marking resolved.

robrap reviewed (commented):
Not quite done with the review, but submitting so you can see comments.

robrap commented on openedx/core/djangoapps/oauth_dispatch/jwt.py:
```diff
@@ -63,6 +66,11 @@ def create_jwt_token_dict(token_dict, oauth_adapter, use_asymmetric_key=None):
+        grant_type = access_token.application.authorization_grant_type
+    except Exception:
+        grant_type = ''
+        log.exception('Unable to get grant_type from access token.')
```
Note: if we don't get this in the logs by the time we are going to merge the final PR, we could remove this try/except.

robrap replied:
I moved this note to my ticket. Marking as resolved.

robrap commented on openedx/core/djangoapps/oauth_dispatch/jwt.py:
```diff
@@ -71,6 +79,7 @@ def create_jwt_token_dict(token_dict, oauth_adapter, use_asymmetric_key=None):
         use_asymmetric_key=use_asymmetric_key,
         is_restricted=oauth_adapter.is_client_restricted(client),
         filters=oauth_adapter.get_authorization_filters(client),
+        grant_type=grant_type,
```
What is this change for?

jawad-khan replied:
As you said I moved this `grant_type = access_token.application.authorization_grant_type` in try catch above and added grant type here.

robrap replied:
Apologies. The way I had github set up it was wrapping strangely, and I thought this was being added as a new parameter to `get_authorization_filters`. My bad.

robrap commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-mobile-migration-to-jwt.rst:
```diff
@@ -0,0 +1,61 @@
+14. Mobile migration to JWT
```
Thanks for this. I'm going to add some edits in a separate commit.

robrap reviewed (commented):
One possibly-blocking comment. Thanks.

robrap reviewed (approved):
@jawad-khan: This is ready from my end. Please review the ADR changes, and squash and merge if you are content. If not, I can review your questions or updates. Thanks.

robrap commented on openedx/core/djangoapps/oauth_dispatch/jwt.py:
```diff
@@ -63,6 +67,15 @@ def create_jwt_token_dict(token_dict, oauth_adapter, use_asymmetric_key=None):
+
+    # .. custom_attribute_name: create_jwt_grant_type
+    # .. custom_attribute_description: The grant type of the newly created JWT.
+    set_custom_attribute('create_jwt_grant_type', grant_type)
```
@jawad-khan: [inform] I added a custom attribute in case we want to review this data in New Relic.

robrap commented on openedx/core/djangoapps/oauth_dispatch/jwt.py:
```diff
@@ -63,6 +67,16 @@ def create_jwt_token_dict(token_dict, oauth_adapter, use_asymmetric_key=None):
+    try:
+        grant_type = access_token.application.authorization_grant_type
+    except Exception:  # pylint: disable=broad-except
+        # TODO: Remove this broad except if proven this doesn't happen.
+        grant_type = 'unknown-error'
```
@jawad-khan: [inform] I changed the grant type for this error condition so we can more easily tell the difference between this issue, and an empty string in the db.

jawad-khan committed:
feat: Added token grant_type in JWT payload

To exchange jwt with session cookies we need to determine JWT grant type in
AccessTokenExchangeView. JWT only having password grant type will be allowed to exchange session.
Added ADR for mobile migration to JWT authentication.

LEARNER-8886

robrap commented:
[inform] I tried rebasing, but still getting test failures. I am trying to re-run to see what happens.
